ospfd/ospf_gr_helper.c

```diff
@@ -0,0 +1,982 @@
+ *
+ * ospf
+ *    ospf pointer
+ * lsa
+ *    topo change occured due to this lsa(type (1-5  and 7)
+ *
+ * Returns:
+ *    Nothing
+ */
+void ospf_helper_handle_topo_chg(struct ospf *ospf, struct ospf_lsa *lsa)
+{
+	struct listnode *node;
+	struct ospf_interface *oi;
+
+	if (!ospf->active_restarter_cnt)
+		return;
+
+	/* Topo change not required to be hanlded if strict
+	 * LSA check is disbaled for this router.
+	 */
+	if (!ospf->strict_lsa_check)
+		return;
+
+	if (IS_DEBUG_OSPF_GR_HELPER)
+		zlog_debug(
+			"%s, Topo change detected due to lsa LSID:%s type:%d",
+			__PRETTY_FUNCTION__, inet_ntoa(lsa->data->id),
+			lsa->data->type);
+
+	lsa->to_be_acknowledged = OSPF_TRUE;
+
+	for (ALL_LIST_ELEMENTS_RO(ospf->oiflist, node, oi)) {
+		struct route_node *rn = NULL;
+
+		if (ospf_interface_neighbor_count(oi) == 0)
+			continue;
+
+		/* Ref rfc3623 section 3.2.3.b
+		 * If change due to external LSA and if the area is
+		 * stub, then it is not a topo change. Since Type-5
+		 * lsas will not be flooded in stub area.
+		 */
+		if ((oi->area->external_routing == OSPF_AREA_STUB)
+		    && (lsa->data->type == OSPF_AS_EXTERNAL_LSA)) {
+			continue;
+		}
+
+		for (rn = route_top(oi->nbrs); rn; rn = route_next(rn)) {
+			struct ospf_neighbor *nbr = NULL;
+
+			if (!rn->info)
+				continue;
+
+			nbr = rn->info;
+
+			if (I_AM_HELPER(nbr))
+				ospf_gr_helper_exit(nbr,
+						    OSPF_GR_HELPER_TOPO_CHG);
+		}
+	}
+}
+
+/*
+ * Api to exit from HELPER role to take all actions
+ * required at exit.
+ * Ref rfc3623 section 3.2
+ *
+ * ospf
+ *    Ospf pointer.
+ *
+ * nbr
+ *    Ospf neighbour for which it is acting as HELPER.
+ *
+ * reason
+ *    The reason for exiting from HELPER.
+ *
+ * Returns:
+ *    Nothing.
+ */
+void ospf_gr_helper_exit(struct ospf_neighbor *nbr,
+			 enum ospf_helper_exit_reason reason)
+{
+	struct ospf_interface *oi = nbr->oi;
+	struct ospf *ospf = oi->ospf;
+
+	if (!I_AM_HELPER(nbr))
+		return;
+
+	if (IS_DEBUG_OSPF_GR_HELPER)
+		zlog_debug("%s, Exiting from HELPER support to %s, due to %s",
+			   __PRETTY_FUNCTION__, inet_ntoa(nbr->src),
+			   ospf_exit_reason_desc[reason]);
+
+	/* Reset helper status*/
+	nbr->gr_helper_info.gr_helper_status = OSPF_GR_NOT_HELPER;
+	nbr->gr_helper_info.helper_exit_reason = reason;
+	nbr->gr_helper_info.actual_grace_period = 0;
+	nbr->gr_helper_info.recvd_grace_period = 0;
+	nbr->gr_helper_info.gr_restart_reason = 0;
+	ospf->last_exit_reason = reason;
+
+	if (ospf->active_restarter_cnt <= 0) {
+		zlog_err(
+			"OSPF GR-Helper: active_restarter_cnt should be greater than zero here.");
+		return;
+	}
+	/* Decrement active Restarter count */
+	ospf->active_restarter_cnt--;
+
+	/* If the exit not triggered due to grace timer
+	 * expairy , stop the grace timer.
+	 */
+	if (reason != OSPF_GR_HELPER_GRACE_TIMEOUT)
+		THREAD_OFF(nbr->gr_helper_info.t_grace_timer);
+
+	/* check exit triggered due to successfull completion
+	 * of gracefull restart.
+	 * If no, bringdown the neighbour.
```

Comment:
what does "bring down" mean here?

Reply:
Here the intend , nbr (RESTARTER)  should be deleted from NBR table and start the nbr FSM from init.